File: clojure/src/aoc/day5.clj
(ns aoc.day5
  (:require [clojure.string :as string]
            [aoc.util :as util]))

(def input "./../inputs/day5.txt")

(defn- parse-seat-id [seat]
  (->>
   seat
   (map #(if (#{\F \L} %) "0" "1"))
   string/join
   (util/str->int 2)))

(defn- read-input []
  (->>
   input
   (slurp)
   (string/split-lines)
   (map parse-seat-id)))

(defn run-part-1 []
  (apply max (read-input)))

(defn run-part-2 []
  (->>
   (read-input)
   (sort)
   (partition 2 1)
   (filter (fn [[a b]] (= a (- b 2))))
   (map #(+ (first %) 1))
   (first)))

File: elixir/lib/aoc/day5.ex
defmodule Aoc.Day5 do
  @input Path.join([__DIR__, "..", "..", "..", "inputs", "day5.txt"])

  def run_part_1 do
    Enum.max(read_input())
  end

  def run_part_2 do
    read_input()
    |> Enum.sort()
    |> Enum.chunk_every(2, 1, :discard)
    |> Enum.filter(fn [a, b] -> a == b - 2 end)
    |> Enum.map(fn [a, _] -> a + 1 end)
    |> List.first()
  end

  defp read_input do
    @input
    |> File.stream!()
    |> Stream.map(&String.trim/1)
    |> Stream.map(fn line ->
      line
      |> String.replace(~r/./, fn
        c when c in ~w[F L] -> "0"
        c when c in ~w[B R] -> "1"
      end)
      |> String.to_integer(2)
    end)
  end
end
